src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -0,0 +1,345 @@
+    if (!ret && (depth == 0 || !m_current_path[depth - 1].depth_exceeded))
+        not_modified.insert(object_key);
+    return ret;
+}
+
+bool DeepChangeChecker::operator()(ObjKeyType key)
+{
+    // If the root object changed we do not need to iterate over every row since a notification needs to be sent
+    // anyway.
+    if (m_root_object_changes &&
+        m_root_object_changes->modifications_contains(key, m_filtered_columns_in_root_table)) {
+        return true;
+    }
+
+    return check_row(m_root_table, key, m_filtered_columns, 0);
+}
+
+KeyPathChangeChecker::KeyPathChangeChecker(TransactionChangeInfo const& info, Table const& root_table,
+                                           std::vector<RelatedTable> const& related_tables,
+                                           std::vector<KeyPathArray> key_path_arrays)
+    : DeepChangeChecker(info, root_table, related_tables, key_path_arrays)
+{
+}
+
+bool KeyPathChangeChecker::operator()(ObjKeyType object_key)
+{
+    // If the root object changed we do not need to iterate over every row since a notification needs to be sent
+    // anyway.
+    if (m_root_object_changes &&
+        m_root_object_changes->modifications_contains(object_key, m_filtered_columns_in_root_table)) {
+        return true;
+    }
+
+    // The `KeyPathChangeChecker` traverses along the given key path arrays and only those to check for changes
+    // along them.
+    for (auto&& key_path_array : m_key_path_arrays) {
+        for (auto&& key_path : key_path_array) {
+            auto next_object_key_to_check = object_key;
+            for (size_t i = 0; i < key_path.size(); i++) {
+                // Check for a change on the current depth.
+                auto table_key = key_path[i].first;
+                auto column_key = key_path[i].second;
+                auto iterator = m_info.tables.find(table_key.value);
+                if (iterator != m_info.tables.end() &&
+                    iterator->second.modifications_contains(next_object_key_to_check, {column_key})) {
+                    return true;
+                }
+
+                // Advance one level deeper into the key path if possible.
+                if (i < key_path.size() - 1) {
+                    auto column_type = column_key.get_type();
+                    if (column_type == col_type_Link) {
+                        const Obj obj = m_root_table.get_object(ObjKey(next_object_key_to_check));
+                        next_object_key_to_check = obj.get<ObjKey>(ColKey(column_key)).value;
+                    }
+                }
+            }
+        }
+    }
+
+    return false;
+}
+
+ObjectChangeChecker::ObjectChangeChecker(TransactionChangeInfo const& info, Table const& root_table,
+                                         std::vector<RelatedTable> const& related_tables,
+                                         std::vector<KeyPathArray> key_path_arrays)
+    : DeepChangeChecker(info, root_table, related_tables, key_path_arrays)
```

Comment:
Same question about `std::move(key_path_arrays)` here to avoid copying arrays of arrays a whole bunch.